### PR TITLE
fault-tolerant: fix resetting "current" progress at the end of successful epoch

### LINKED ISSUE #8835: resume_from_checkpoint broken when fault-tolerant feature enabled
## 🐛 Bug

A trainer with the resume from checkpoint option does not continue training and stops immediately, despite increased max_epoch settings.

### To Reproduce
```python

import os
import unittest.mock

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        print(batch.sum())
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def on_train_epoch_end(self):
        print("epoch ended")

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

    def on_load_checkpoint(self, checkpoint):
        pass


@unittest.mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=3,
        limit_val_batches=0,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
    )
    trainer.fit(model, train_dataloader=train_data)

    trainer.save_checkpoint("lightning_logs/auto.pt")

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=3,
        limit_val_batches=0,
        num_sanity_val_steps=0,
        max_epochs=3,
        weights_summary=None,
        resume_from_checkpoint="lightning_logs/auto.pt",
    )
    trainer.fit(model, train_dataloader=train_data)


if __name__ == "__main__":
    run()


```

Output: 

```python
Epoch 0:   0%|          | 0/3 [00:00<?, ?it/s] tensor(6.2218)
Epoch 0:  33%|███▎      | 1/3 [00:00<00:00, 83.40it/s, loss=-1.23, v_num=82]tensor(-9.5560)
Epoch 0:  67%|██████▋   | 2/3 [00:00<00:00, 120.73it/s, loss=-1.71, v_num=82]tensor(14.5846)
Epoch 0: 100%|██████████| 3/3 [00:00<00:00, 140.56it/s, loss=-0.136, v_num=82]epoch ended
Epoch 0: 100%|██████████| 3/3 [00:00<00:00, 98.17it/s, loss=-0.136, v_num=82] 

Restoring states from the checkpoint path at lightning_logs/auto.pt
Epoch 1: 100%|██████████| 3/3 [00:00<00:00, 22712.84it/s]epoch ended
Epoch 2: 100%|██████████| 3/3 [00:00<00:00, 20004.63it/s]epoch ended
Epoch 2: 100%|██████████| 3/3 [00:00<00:00, 3038.62it/s] 
```
Here the training_step does not get invoked on the second fit call. Instead, the epochs end immediately.

ONLY happens when PL_FAULT_TOLERANT_TRAINING=1. This is an experimental feature which is off by default.

### Expected behavior

The training continues due to the increased max_epochs setting.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8835

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -229,6 +229,10 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
 
         self.update_lr_schedulers("epoch", update_plateau_schedulers=True)
 
+        self.batch_progress.current.reset()
+        self.scheduler_progress.current.reset()
+        self.batch_loop.optim_progress.reset_on_epoch()
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:234]: same as in reset() above line 90-93

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:234]: Notice that it's calling `reset_on_epoch`, not `reset`. This loop is the epoch loop

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:234]: Probably worth extracting into a protected method

**Review by carmocca — COMMENTED** [review]
Does the bug happen because `restarting=False` and the `current` values correspond to those of the  finished epoch? So we never roll over to the next one?
